### PR TITLE
Support multiple source roots and namespace packages

### PR DESCRIPTION
**Pull request by emdoyle**
Fixing #203 and #192 

NOTE: The base of this branch will be updated after #207 merges

This PR essentially changes `source_root` from a single path to a list of paths (`source_roots`).
This change propagates through most of the codebase, but the semantic change is straightforward.

Tach will now search for project imports in _all_ the configured source roots when translating module paths to filesystem paths (and vice versa).

This unlocks support for 'monorepo' use cases, since a user can now mark all applicable package source roots and validate the expected dependencies.

This PR also drops certain expectations around finding `__init__.py` in Python packages, which means [namespace packages](https://peps.python.org/pep-0420/) are also supported. This is particularly useful when building many subsidiary packages which are distributed under a common namespace.

Example:
image

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -147,6 +147,7 @@ python/tach/cache/setup.py[L7]: Import 'tach.filesystem.find_project_config_root
 Tach also supports:
 
 - [Manual file configuration](https://docs.gauge.sh/usage/configuration)
+- [Monorepos and namespace packages](https://docs.gauge.sh/usage/configuration#source-roots)
```
Comment (caelean) [inline README.md:150]: Small nit - thoughts on the ordering of the list here? Was trying to go from most important to least for the average user, almost put deprecations under inline exceptions. Imo this is prob under strict interfaces/maybe we want to move manual file configuration down too

File: docs/usage/configuration.mdx [hunk 2]
```diff
@@ -57,48 +57,78 @@ Each module listed under the `modules` key above can accept the following attrib
...
-This means Tach will expect that your Python imports are resolved relative to the directory in which `tach.yml` exists (the 'project root').
+The `source_roots` key is required for Tach to understand the imports within your project.
+If it is not set explicitly, `source_roots` defaults to your project root path: `['.']`.
+This means Tach will expect that your Python imports are resolved relative to the directory in which `tach.yml` exists.
 
-Below is a typical case in which `source_root` is necessary.
+Below is a typical case in which modifying `source_roots` is necessary.
 
-### Example
+### Example: Monorepo
```
Comment (caelean) [inline docs/usage/configuration.mdx:70]: Should this replace the standard example, or should this be in addition to it? I could see monorepo/multiple source root configuration being it's own docs page as it's less likely to represent the path that most take

File: python/tach/core/config.py [hunk 3]
```diff
@@ -83,15 +83,15 @@ class ProjectConfig(Config):
...
     exact: bool = False
     disable_logging: bool = False
     ignore_type_checking_imports: bool = True
     forbid_circular_dependencies: bool = False
 
-    @field_serializer("source_root")
-    def serialize_source_root(self, source_root: Path, _) -> str:
-        return str(source_root)
+    @field_serializer("source_roots")
+    def serialize_source_root(self, source_roots: list[Path], _) -> list[str]:
```
Comment (caelean) [inline python/tach/core/config.py]:
```suggestion
    def serialize_source_roots(self, source_roots: list[Path], _) -> list[str]:
```

File: python/tach/mod.py [hunk 4]
```diff
@@ -56,11 +60,14 @@ def validate_configuration(
...
-        if configuration.source_root not in module_path.parents:
+        if not any(
+            source_root in module_path.parents
+            for source_root in configuration.source_roots
+        ):
             # This module exists outside of the source root
             # This is not allowed and should be reported as a configuration error
             errors.append(
-                f"Module '{module_path}' is not contained within source root: '{configuration.source_root}'"
+                f"Module '{module_path}' is not contained within any source root: {configuration.source_roots}"
```
Comment (caelean) [inline python/tach/mod.py:70]:
```suggestion
                f"Module '{module_path}' is not contained within any source root: {[str(root) for root in configuration.source_roots]}"
```

**Review by caelean — APPROVED** [review]

**Comment by caelean** [comment]
Looks great/working for me when I test! Some nice-to-haves would mainly be around the ergonomics of `mod` - unmarking modules that don't have a source root, not allowing nested source roots, not allowing modules with the same relative path under different source roots, etc. 

Definitely all nonblocking tho; stoked to see this get out!

File: README.md (hunk 1 above)
Comment (emdoyle) [reply README.md]: Updated this :+1:

File: docs/usage/configuration.mdx (hunk 2 above)
Comment (emdoyle) [reply docs/usage/configuration.mdx:109]: Good idea, restored the previous example

**Comment by malcolmgreaves** [comment]
This is awesome! Just tried this today on my own project and it worked so well! Thank you so much :) ❤️

**Pull request merged**